### PR TITLE
Refactor and add do resize async in renderQueue

### PR DESCRIPTION
**Pull request by megheaiulian**
Refactor and add do resize async in renderQueue

### CODE REVIEW

File: cosmoz-data-nav.js [hunk 1]
```diff
@@ -337,13 +337,14 @@
...
 			this._preload();
 
-			if (this.animating) {
+			if (this.animating || this.selected == null) {
 				return;
 			}
-			if (index === this.selected) {
-				return this._updateSelected();
+			if (this.selected !== index) {
+				this._synchronize();
```
Comment (nomego) [inline cosmoz-data-nav.js]: _updateSelected will run _synchronize, isn't that enough?

File: cosmoz-data-nav.js [hunk 2]
```diff
@@ -488,52 +483,38 @@
...
-		 * @return {void}
-		 */
-		_forwardItem(element, item) {
-			if (element.item === item) {
-				// already rendered
+			if (!this.isIncompleteFn(item) && element.item === item) {
 				return;
 			}
+			element._reset = true;
+			console.log('resetting item', index);
```
Comment (nomego) [inline cosmoz-data-nav.js]: no logs

File: cosmoz-data-nav.js [hunk 3]
```diff
@@ -730,17 +723,29 @@
 				return;
 			}
 
+			const selected = this.selected;
```
Comment (nomego) [inline cosmoz-data-nav.js]: Why is this chunk better than sort? To me, it's harder to read and more error-prone to maintain?

File: cosmoz-data-nav.js [hunk 4]
```diff
@@ -730,17 +723,29 @@
...
 
 			this._indexRenderQueue = queue
-				.sort((a, b) => a === this.selected ? -1 : b === this.selected ? 1 : 0)
 				.map(idx => {
 
 					const elementIndex = idx % this.elementsBuffer,
 						element = this._elements[elementIndex],
 						item = this.items[idx];
 
+					if (renderRun) {
```
Comment (nomego) [inline cosmoz-data-nav.js]: Why is this moved so early in the queue mgmt? I think it would make more sense to put it right before _forwardItem to just avoid the expensive templating but still be able to evaluate reRun or incomplete templates the same run.

File: cosmoz-data-nav.js [hunk 5]
```diff
@@ -749,29 +754,31 @@
...
+						element.item = false;
+						// no data for item drop task from queue
 						return;
 					}
 
+					element.__incomplete._showHideChildren(true);
+
 					if (element.item === item) {
-						// already rendered
+						renderRun = idx === this.selected && this._notifyElementResize();
```
Comment (nomego) [inline cosmoz-data-nav.js]: This adds a dependency to https://github.com/Neovici/cosmoz-data-nav/pull/21 and `renderRun` should be Boolean.

File: cosmoz-data-nav.js [hunk 6]
```diff
@@ -749,29 +754,31 @@
...
 						// drop task from queue
 						return;
 					}
 
-					if (renderRun) {
-						// one render per run
-						// maintain task in queue
+					this._forwardItem(element, item, idx);
+
+					renderRun = idx !== this.selected;
```
Comment (nomego) [inline cosmoz-data-nav.js]: `renderRun` is used for multiple things? This can make two renders occur during the same run?
One thing is to leave the index for resizing next round (which shouldn't be needed if it only applies to selected) and another thing is to decide if we want to render more.

File: cosmoz-data-nav.js [hunk 7]
```diff
@@ -488,52 +483,38 @@
...
+			if (!this.isIncompleteFn(item) && element.item === item) {
 				return;
 			}
+			element._reset = true;
+			console.log('resetting item', index);
 
-			this._removeInstance(element.__instance);
+			const baseProps = this._getBaseProps(index),
+				incomplete = element.__incomplete,
+				instance = element.__instance;
```
Comment (nomego) [inline cosmoz-data-nav.js]: Since resetElement only will be run once per element, we shouldn't do any work on instance, only incomplete.

**Review by nomego — CHANGES_REQUESTED** [review]
The start of the PR has some improvements to flow, simplifications, helpers and possible perf improvements, but I think the changes to toggle/reset/renderQueue has to be motivated further because I can't see what they improve at this point. Also I'm not sure notifyElementResize should be a part of the queue this way or rather be described through its own dependencies and conditions.

File: cosmoz-data-nav.js (hunk 1 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: This can be refactored a bit better.

File: cosmoz-data-nav.js (hunk 7 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: Element can be complete but have a different old item and needs to be in reset state until it will be templatized again.

File: cosmoz-data-nav.js (hunk 3 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: The selected element should be moved to first position in the Array. I don't see why sort would be used instead of simply removing it from where it is and putting it to first position.

File: cosmoz-data-nav.js (hunk 4 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: We should not do anything else after a render has run. Incomplete items might be complete in next run.

File: cosmoz-data-nav.js (hunk 6 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: If the newly created element is for the currently selected instance we must resize it. That call will be expensive so we keep the current index in the array and next call will resize instance.

File: cosmoz-data-nav.js (hunk 5 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: Yes the PR should be merged after #21. Will change it so that `renderRun` will be  a Boolean. 
If `idx` is the selected index and we have done a element resize we should not do more in this run because the resize has been expensive.

File: cosmoz-data-nav.js (hunk 7 above)
Comment (nomego) [reply cosmoz-data-nav.js]: Well if `element._reset` is true, the `_resetElement()` will never do any work on that element again?

File: cosmoz-data-nav.js (hunk 3 above)
Comment (nomego) [reply cosmoz-data-nav.js]: Well mostly I think it hurts readability, and extensibility.

There are now 7 lines to read and understand as opposed to one. 
You can't skip reading the block because it's not clear what it does until you've read it all.
You would also slow down to evaluate why selected = 0 or selectedIndex = 0 wouldn't apply for this code.
It would also fail on corner-cases like selected = 0 but not first task in queue.

I'm not sure this is more performant, and if it is, it wouldn't matter for a Number array of 3-10 items.

It will be messy to extend, if we, for example, want to sort the queue on more than one dimension. Selected, needs resize, etc, etc.

File: cosmoz-data-nav.js (hunk 4 above)
Comment (nomego) [reply cosmoz-data-nav.js]: I disagree, I think we should do whatever inexpensive task that is left in the queue but avoid the expensive ones.
Since `resetElement` won't reset a previously reset element, we should make sure that the elements in the buffer are in sync in case the user navigates back or forth to show the loading, since we won't render during animation.

File: cosmoz-data-nav.js (hunk 6 above)
Comment (nomego) [reply cosmoz-data-nav.js]: Ok then I think this would be clearer by just changing the if statement to `if (idx === this.selected)` instead of double negatives and re-using renderRun.
Otherwise you would actually render selected and set renderRun to `false`, causing it to possibly run twice.

**Review by nomego — CHANGES_REQUESTED** [review]

File: cosmoz-data-nav.js (hunk 4 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: @nomego We set `renderRun` to `true` manually so that it returns the idx for the next items. If we want something more to run we should not set renderRun to true?

Comment (nomego) [reply cosmoz-data-nav.js]: Well it only prevents any _render_ task to run, it could still empty the queue of tasks setting incompelte status/toggling instances.

File: cosmoz-data-nav.js (hunk 3 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: Will revert this to sort.

File: cosmoz-data-nav.js (hunk 7 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: Yes. I am not filtering in synchronize because _resetElement will check that.

File: cosmoz-data-nav.js (hunk 3 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js]: @nomego Reverted to sort.
